packages/backend-common/src/scm/git.ts

```diff
@@ -0,0 +1,209 @@
+ */
+import git, { ProgressCallback } from 'isomorphic-git';
+import http from 'isomorphic-git/http/node';
+import fs from 'fs-extra';
+import { Logger } from 'winston';
+
+/*
+provider    username  	    password
+GitHub      token	          'x-oauth-basic'
+GitHub App 	token	          'x-access-token'
+BitBucket	  'x-token-auth'	token
+GitLab      'oauth2'	      token
+From : https://isomorphic-git.org/docs/en/onAuth
+
+Azure       'notempty'      token
+*/
+class SCM {
+  constructor(
+    private readonly config: {
+      username?: string;
+      password?: string;
+      logger?: Logger;
+    },
+  ) {}
+
+  async add({ dir, filepath }: { dir: string; filepath: string }) {
+    this.config.logger?.info(`Adding file {dir=${dir},filepath=${filepath}}`);
+
+    return git.add({ fs, dir, filepath });
+  }
+
+  async addRemote({
+    dir,
+    url,
+    remoteName,
+  }: {
+    dir: string;
+    remoteName: string;
+    url: string;
+  }) {
+    this.config.logger?.info(
+      `Creating new remote {dir=${dir},remoteName=${remoteName},url=${url}}`,
+    );
+    return git.addRemote({ fs, dir, remote: remoteName, url });
+  }
+
+  async commit({
+    dir,
+    message,
+    author,
+    committer,
+  }: {
+    dir: string;
+    message: string;
+    author: { name: string; email: string };
+    committer: { name: string; email: string };
+  }) {
+    this.config.logger?.info(
+      `Committing file to repo {dir=${dir},message=${message}}`,
+    );
+
+    return git.commit({ fs, dir, message, author, committer });
+  }
+
+  async clone({ url, dir }: { url: string; dir: string }) {
+    this.config.logger?.info(`Cloning repo {dir=${dir},url=${url}}`);
+    return git.clone({
+      fs,
+      http,
+      url,
+      dir,
+      singleBranch: true,
+      depth: 1,
+      onProgress: this.onProgressHandler,
+      headers: {
+        'user-agent': 'git/@isomorphic-git',
+      },
+      onAuth: this.onAuth,
+    });
+  }
+
+  // https://isomorphic-git.org/docs/en/currentBranch
+  async currentBranch({ dir, fullName }: { dir: string; fullName?: boolean }) {
+    const fullname = fullName ?? false;
+    return git.currentBranch({ fs, dir, fullname });
+  }
+
+  // https://isomorphic-git.org/docs/en/fetch
+  async fetch({ dir, remote }: { dir: string; remote?: string }) {
+    const remoteValue = remote ?? 'origin';
+    this.config.logger?.info(
+      `Fetching remote=${remoteValue} for repository {dir=${dir}}`,
+    );
+    return git.fetch({
+      fs,
+      http,
+      dir,
+      remote: remoteValue,
+      onProgress: this.onProgressHandler,
+      headers: {
+        'user-agent': 'git/@isomorphic-git',
+      },
+      onAuth: this.onAuth,
+    });
+  }
+
+  async init({ dir }: { dir: string }) {
+    this.config.logger?.info(`Init git repository {dir=${dir}}`);
+
+    return git.init({
+      fs,
+      dir,
+    });
+  }
+
+  // https://isomorphic-git.org/docs/en/merge
+  async merge({
+    dir,
+    headBranch,
```

Comment:
perhaps just keep `ours` and `theirs` here? They aren't necessarily branches either right?

Reply:
Yeah I guess this is only for branches. Documentation says `Merge two branches`. Or at least that is what we are using this `merge` for. Found `head` and `base` easier to understand.

Reply:
Maybe we can change it to `headRef` and `baseRef` instead?

Reply:
Merges are only for branches as far as I know? *googles furiously*

Reply:
Implementation of the thing is looking like you can pass any kind of ref, including sha. Main concern was that I'm thinking it's a bit unnecessary to rename the parameters at all compared to isomorphic-git, unless we actually change their functionality. Same goes for `remoteName` above, feeling it's better to just stick to the existing names to make it easier to follow the documentation